src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/RestoreArgs.cs

```diff
@@ -109,52 +109,33 @@ public List<SourceRepository> GetEffectiveSources(ISettings settings)
 
         private List<SourceRepository> GetEffectiveSourcesCore(ISettings settings)
         {
-            // Take the passed in sources
-            var packageSources = new HashSet<string>(Sources, StringComparer.Ordinal);
             var sourceObjects = new Dictionary<string, PackageSource>(StringComparer.Ordinal);
+            var packageSourceProvider = new PackageSourceProvider(settings);
+            var packageSourcesFromProvider = packageSourceProvider.LoadPackageSources();
+            var useNugetConfigSources = (Sources.Count == 0);
 
-            var packageSourceProvider = new Lazy<PackageSourceProvider>(() 
-                => new PackageSourceProvider(settings));
+            // Always use passed-in sources and fallback sources
+            foreach (var sourceUri in Enumerable.Concat(Sources, FallbackSources))
+            {
+                sourceObjects[sourceUri] = new PackageSource(sourceUri);
+            }
 
-            // If no sources were passed in use the NuGet.Config sources
-            if (packageSources.Count < 1)
+            // Use PackageSource objects from the provider when possible (since those will have credentials from nuget.config)
+            foreach (var source in packageSourcesFromProvider)
             {
-                // Add enabled sources
-                foreach (var source in packageSourceProvider.Value.LoadPackageSources())
+                if (source.IsEnabled && (useNugetConfigSources || sourceObjects.ContainsKey(source.Source)))
```

Comment:
There is a test for that, these new XPlat functional tests specify a source on the command line and expect the credentials to come from nuget.config.
There should be no change in behavior here besides fixing the credential bug. If the user passes in a source on the command line, only that will be added to sourceObjects.

---
In reply to: [64119109](https://github.com/NuGet/NuGet.Client/pull/616#discussion_r64119109) [](ancestors = 64119109)